libgedathon/docs/gsym-attribute.1.in
.TH gsym-attribute 1 "@DATE@" "gEDA Project" @VERSION@

.SH NAME
gsym-attribute \- gEDA-gaf Symbol File Attribute Manipulator Utility

.SH SYNOPSIS
.B gsym-attribute
.RI  [ INFORMATION ] | [ OPTIONS ] [ DIRECTIVES ] [INPUT] [OUTPUT]

.SH DESCRIPTION
.B gsym-attribute
Sets or retrieve the value of a Text attribute in a symbol file

.SH QUICK
.B usage: [-h] | [Options] [-a] attribute -i <inputfile> [[-o] <outputfile> ]

ex. 1) Set the footprint in all symbol files in the current directory to SO8:

            gsym-attribute.py -a footprint -l SO8

ex. 2) Set the author in files in 24LC16BIMC-1.sym 24LC16BEMC-1.sym to Jethro Bodine

            gsym-attribute.py -v -a author -l "Jethro Bodine" 24LC16BIMC-1 24LC16BEMC-1

ex. 3) Get the value of slotdef in the file dual-opamp-3.sym

            gsym-attribute.py slotdef dual-opamp-3

.SH INFORMATION
Use of the following options displays information related to gsym-attribute
but does not inspect or alter any symbol files:
.IP "-h, --help"
Show information about the advanced usage of this command.
.IP "-V, --version"
Displays software version information.

.SH OPTIONS
.IP "-l, --value"
Optional new value for the attribute.
.IP "-c, --color"
Optional attribute color.
.IP "-s, --size"
Optional font size.

.SH DIRECTIVES
.IP "-f, --force"
When the force option is specified existing file matching the output name will be over-written, without any warning.
.IP "-a, --attribute"
The Attribute to be queried or set.
.IP "-d, --hidden"
When creating an attribute, set the visibility flag to invisible. If the -l option is used and a matching attribute
is not found, a new attribute is automatically created. By default new attributes are visible, use this option to make
the new attribute invisible.
.IP "-q, --quite"
Quite mode. Suppress normal processing messages.
.IP "-v, --verbose"
Enable verbose mode. Displays additional information about what the program is doing.
.IP "-i, --input <input file>"
Specify the name of the input sym file. Wild cards are accepted, the default is *.sym.
.IP "-o, --output <output file>"
Specifies the name of the output sym file. When the output file name is not specified the input file will be over written

.SH HOMEPAGE
http://www.geda-project.org/

.SH BUGS
You can report bugs at the Launchpad site of gsym-attribute:
https://bugs.launchpad.net/geda/gsym-attribute/+filebug

.SH AUTHOR
Wiley Edward Hill < wileyhill dot gmail dot com >

.SH COPYRIGHT
.nf
Copyright \(co @YEAR@ Wiley Edward Hill.  License GPLv3+: GNU GPL
version 3.  Please see the `COPYING' file included with this
program for full details.
.PP
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

.SH SEE ALSO
.BR gsym-set-font-size (1) gsym-set-visibility (1) geda-snap-grid (1)

.


NEWS
==============================================
 'gschem and Friends' Electronic Design Suite
==============================================

Copyright (C) 1998-2017 gEDA Developers

This file documents important user-visible changes in gEDA/gaf.  For
more information, please consult the `ChangeLog' file.

Notable changes in gEDA/gaf gEDA 2.1.1.1 - 11/12/17 Rampage Edition

* Expanded support for internationalization (i18n).

* The build system now support using the --with-included-gettext option to
  configure the suite to build and use an in-source international library.

* Font-sizes for widgets on GedaToolbars can now be specified in RC files.

* The toolbar layout in gschem has been improved to better support labels.

* The visibility of toolbar tooltips can now be control using RC keywords.

* The Slot Editor Dialog now incorporates a spinner control to adjust the
  slot number of slotted components. The range is restricted to valid slots
  for the selected component.

* New bitmaps have been added;

      * geda-deslect.png
      * geda-find-attribute
      * geda-properties.png
      * geda-trash-can.png

* libgedathon no longer over-writes existing gafrc when the declare_local_sym
  feature is used.

* libgedathon now supports internationalization (i18n).

Notable changes in gEDA/gaf gEDA 2.1.1.0 - 01/01/17 Rampage Edition

* Improvements were made to both the build system and source code to ensure
  gEDA-GAF builds more reliably on 64 bit platforms. Testing was performed
  using Ubuntu 12.04 and Ubuntu 14.04 64 bit host systems.

* The libgeda API has been completely redone in order to present a library API
  that is not only for in-suite applications.

* Support for internationalization (i18n) has been significantly expanded and
  now includes command-line utilities.

* Extensive revisions have been made to improve performance of gEDA application.
  This is particularly noticeable in gschem.

* A new library, libgedacolor, now unifies color management for the gEDA suite.
  Color-maps are still configured using scheme files, but the use of "friendly-
  color-maps"  is no longer applicable; the new color maps for libgedacolor are
  already "friendly". The library was added to provide a unified API for gEDA
  applications and should be transparent to most users. User who are using
  custom color-maps will need to modify their custom map files.

* A new auto-panning feature has been added to gschem. The auto-pan is only in
  effect while drawing nets and buses and can enabled or disabled through menu
  options.

* The Auto-Number dialog in gschem no longer requires the use of wild-card
  characters in the scope text. This is a new advanced feature that allows
  automatic re-numbering of duplicated reference designators.

* The path for logs file can now be set in initialization files, including
  local gafrc files. This allows logs to be organized on a per project basis.
  The location for logs can also be set using a new GEDALOGS environment variable.

* A new libgeda keyword "log-time" can be use to enable or disable prefixing
  log entries with the time of day.

* The Multi-Attribute Editor Dialog entry now has attribute "word" completion
  identical to the Single-Attribute Editor Dialog.

* The list of valid attribute used by gsymcheck when checking symbol files is
  now users configurable. The list can be set in system-gsymcheckrc.

* The list of known devices recognized by gsymcheck when checking the device
  entries in symbol files is now users configurable. The list can be set in
  system-gsymcheckrc.

* Symbol files can now be renamed using the Component Dialog.

* The number of user defined colors was increased by 6 to a total of
  10 user definable colors in each color map.

* The main menu system has been modified to automatically build a basic
  fall-back menu when errors are discovered in the menu RC file.

* The height of the status bar in Gschem can now be set in the gschem-gtkrc

* HotKey dialog now includes actions icon and a column to display the actual
  binding string. User can enable or disable viewing the action column with
  the setting being retained between session.

* The Slot Editor Dialog now displays the total number of slots per component.

* New features have been added to the Page Select Dialog:

      Accept drag-and-drop of schematic or symbol files.
      Opened pages can be re-ordered either up or down.
      Auto scrolls to selection.
      Shows long filename as tooltip when showing short names.
      Optionally adjust height of dialog automatically.

* Exchange image file now retains the user preference to update all selected
  images between session. Previously the user preference was only retained for
  current session.

* Additional preference variables have been incorporated in new the configuration
  system allow retention of changes using the Preferences dialog:

      continue-component-place
      embed-components
      enforce-hierarchy
      force-boundingbox
      keyboardpan-gain
      netconn-rubberband
      select-slack-pixels
      snap-size
      sort-component-library
      object-clipping

  Uses of these keywords in rc files is still permissible and will over-ride the
  preference setting.

* A new gschem command line options has been provided to specify a title-block at
  startup.

* A new gschem command line options has been provided to startup minimized.

* The command line option to add a directory to the Guile path now has a long option.
  The "-L" option has been replaced with "-g" or "--guile".

* gsymcheck now supports the standard command-line version options.

* Programs in the geda-gaf suite now support quite mode in conjunction with the version
  option to assist with extraction of dotted version information.

* New Scheme API have been added:

      gschem-bmp-image
      gschem-jpeg-image
      gschem-tiff-image

* The maximum zoom magnification in gschem has been increased slightly.


libgedathon/scripts/gsym-attribute.py
#! /usr/bin/env python

# Set Visibility of Attributes (For now just to Defaults)

import os, sys, re, getopt, glob
from geda import geda
from geda.constants import *

#---------------------------------------------------------------------------
Version="0.2.2"
#---------------------------------------------------------------------------
VerboseMode=False
#---------------------------------------------------------------------------

#  This is the help string.
Usage =\
"""
gsym-attribute -- A gEDA-gaf Symbol File Attribute Manipulator Utility
                  to set or retrieve the value of a Text attribute in a
                  symbol file

Usage: gsym-attribute [Options] [-a] attribute -i <inputfile> [[-o] <outputfile> ]
"""

Help =\
"""
Options:

  -a, --attribute -- The Attribute to be queried or set
  -u, --value     -- Optional new value for the attribute
  -f, --force     -- When the force option is specified any existing file matching
                     the output name will be over-written, without any warning.
  -c, --color     -- Optional attribute color.
  -d, --hidden    -- Set the visibility flag to invisible When adding an attribute.
  -s, --size      -- Optional font size.
  -q, --quite     -- Suppress normal processing messages.
  -v, --verbose   -- Spews lots of info about what the prog is doing.

  -i, --input  <FILE> -- Specifies the name of the input sym file.

  -o, --output <FILE> -- Specifies the name of the output sym file. When the
                         output file name is not specified the input file will
                         be over written.

Example: Set the footprint in all symbol files in the current directory to SO8:

	        gsym-attribute.py -a footprint -u SO8

Example: Set the author in files in 24LC16BIMC-1.sym 24LC16BEMC-1.sym to Jethro Bodine

            gsym-attribute.py -v -a author -u "Jethro Bodine" 24LC16BIMC-1 24LC16BEMC-1

Example: Get the value of slotdef in the file dual-opamp-3.sym

	        gsym-attribute.py slotdef dual-opamp-3

Copyright (C) 2014-2015 by Wiley Edward Hill.  Released under GPL Version 2.

"""
#------------------------ Functions used by Classes ------------------------

def VMessage(Options, String):
    """
    This prints out String when the -v flag is set, otherwise is silent
    """
    if (Options.VerboseMode == "verbose"):
        print(String)
    return

#---------------------------------------------------------------------------

def QMessage(Options, String):
    """
    This prints out String when the -v flag is set, otherwise is silent
    """
    if Options.VerboseMode == "verbose" or not Options.QuiteMode:
        print(String)
    return

#---------------------------------------------------------------------------

def CheckFilename(Filename, suffix):
    """
    This checks a string to make sure that it is a valid filename.
    It currently doesn't do very much. . . .
    """
    if (re.search('\.' + suffix + '$', Filename)):
        return 1
    else:
        return 0

#-------------------------------- Classes ----------------------------------
class ProgramParameters:

    """
    This class holds info about the environment and cmd line args passed to the
    program.  It has only one method:  the constructor, which gets the args
    and fills out the public vars.  The public vars are:
    """
    def __init__(self):

        global Version

        valid_attributes =[ "device", "footprint", "numslots", "refdes",
                            "author", "dist-license", "use-license",
                            "description", "documentation", "symversion",
                            "graphical", "net", "netname", "slot", "slotdef",
                            "value", "comment", "footprints", "model-name",
                            "file", "pins", "spice-type"]
        """
        Constructor: parse through cmd line args and fill out vars. The values
        here are the defaults
        """
        self.ExitCode         = 0
        self.ForceMode        = False
        self.RecursiveMode    = False
        self.VerboseMode      = False
        self.QuiteMode        = False
        self.InputFiles       = []
        self.UserDir          = os.path.abspath(os.getcwd())
        self.InputFileName    = ""
        self.OutputFileName   = ""

        self.AttributeName    = ""
        self.AttributeValue   = ""
        self.AttributeColor   = 0
        self.AttributeSize    = None
        self.AttributeVisible = VISIBLE;

        #  Get ScratchDir, either from environment, or just use /tmp as default.
        for EnvVar in ["TMP", "TMPVAR", "TEMP"]:
            try:
                TempDir = os.environ[EnvVar]
            except:
                continue                      # Not present, continue looping
            else:
                self.ScratchDir = TempDir     # Got it!
                break
        else:
            self.ScratchDir = "/tmp"          # no env var set, use default

        #  Get and process command line args
        try:
            long_opt = ["force",
                        "help",
                        "attribute=",
                        "color=",
                        "hidden",
                        "input=",
                        "output=",
                        "size=",
                        "value=",
                        "quite",
                        "verbose",
                        "version"]
            OptList, Args = getopt.getopt(sys.argv[1:], 'fha:c:di:o:s:l:qvV', long_opt)
        except getopt.error:
            print Usage                # print out usage string if
                                       # user uses invalid flag.
            sys.exit(1)

        # First pass through args.  Get switch settings & set program modes.
        for Option, Value in OptList:

            if Option in ('-f', '--force'):
                self.ForceMode = True

            if Option in ('-h', '--help'):
                print Usage
                print Help
                sys.exit(0)

            if Option in ('-a', '--attribute'):
                self.AttributeName       = Value

            if Option in ('-c', '--color'):
                self.AttributeColor      = Value

            if Option in ('-d', '--hidden'):
                self.AttributeVisible    = INVISIBLE

            if Option in ('-i', '--input'):
                self.InputFileName       = Value


            if Option in ('-l', '--value'):
                self.AttributeValue      = Value

            if Option in ('-q', '--quite'):
                self.QuiteMode           = True

            if Option in ('-s', '--size'):
                self.AttributeSize       = Value

            if Option in ('-v', '--verbose'):
                self.VerboseMode         = True

            if Option in ('-V', '--version'):
                print Version
                sys.exit(0)

            #if os.path.isfile(Value):
            #    self.InputFiles.append(Value)

        # Second pass.  Do sanity checking and get configured filenames.
        for Option, Value in OptList:
            if Option in ('-o', '--output'):
                if CheckFilename(Value, "sym"):
                    self.OutputFileName = Value
                else:
                    print("Warning -- output file suffix is not \".sym\"")
                    Input = raw_input("Continue? [y/N] ")
                    if ((len(Input) == 0) or (Input[0] != "y") ):
                        sys.exit(1)
                    elif not self.ForceMode and os.path.isfile(Value):
                        print("Warning -- output file exist, continue will over write the file")
                        Input = raw_input("Continue? [y/N] ")
                        if ((len(Input) == 0) or (Input[0] != "y") ):
                            sys.exit(1)
                        else:
                            self.OutputFileName = Value
                    else:
                        self.OutputFileName = Value

        # Third step: Create list of files remaining on command line, and create output
        # base file name.
        self.CmdLineFileList = Args

        # Forth step: If the attribute name was not specified look for a valid name one in
        # the CmdLineFileList list and remove from list and use for the attribute name
        #
        if not self.AttributeName and len(self.CmdLineFileList) > 0:
            string = self.CmdLineFileList[0]
            if string:
                for attribute in valid_attributes:
                    if string == attribute:
                        self.AttributeName = string
                        self.CmdLineFileList.pop(0)
                        break

        self.OutputFileNameBase = re.sub('\.sym', '', self.OutputFileName)

        return

#---------------------------------------------------------------------------

def ProcessSymbol(Options, File):

    symbol = geda.open_page(File)

    if Options.VerboseMode:
        print "Processing: " + symbol.filename

    objects   = geda.get_objects(symbol)
    modified  = 0
    Attribute = None
    """
      Loop thru all objects and look for the attribute
    """
    for capsule in objects:
        object = geda.get_object(capsule)          # Retrieve object from GedaCapsule
        if capsule.type == OBJ_TEXT:
            string = object.string
            index = string.find("=")
            if index > 0:
                if object.name() == Options.AttributeName:
                    Attribute = object
                    break

    if not Attribute:                  # If not found and have value then create new
        if Options.AttributeValue:
            Attribute = geda.new_attrib(Options.AttributeName,
                                        Options.AttributeValue, 100, 100,
                                        Options.AttributeVisible, SHOW_NAME_VALUE)
            geda.add_object(symbol, Attribute)
            modified = 1
        else:                          # No value specified and attribute not found
            QMessage (Options, (Options.AttributeName + ' was not found in ' + symbol.filename))
            Options.ExitCode = 1;
    else:                              # Found the attribute
        if Options.AttributeValue:     # Check the value and set only if different
            if not Attribute.value == Options.AttributeValue:
                string = Options.AttributeName + '=' + Options.AttributeValue
                Attribute.string = string
                modified = 1
        else:                          # Get print the current value
            if Options.VerboseMode:
                print symbol.filename + '::' + Attribute.string
            else:
                print Attribute.string

    if Attribute and Options.AttributeSize:
        Attribute.size = Options.AttributeSize
        modified = 1

    if modified:
        QMessage (Options, ("saving file:" + symbol.filename))
        symbol.save() # Write the symbol to storage

    # Close the file
    symbol.close()

#---------------------------------------------------------------------------

def RecursiveGlob(path, *exts):
    """ Glob recursively a directory and all subdirectories for multiple file extensions
    Note: Glob is case-insensitive, i. e. for '\*.jpg' you will get files ending
    with .jpg and .JPG

    Parameters
    ----------
    path : str
        A directory name
    exts : tuple
        File extensions to glob for

    Returns
    -------
    files : list
        list of files matching extensions in exts in path and subfolders
    """
    dirs = [a[0] for a in os.walk(path)]
    f_filter = [d+e for d in dirs for e in exts]
    return [f for files in [glob.iglob(files) for files in f_filter] for f in files]

#---------------------------------------------------------------------------
#  Main prog begins here

def main(argv):

    foundfiles = False
    foundOurSelf = False

    Options = ProgramParameters()  #  Creates Options object for program varibles.

    if not Options.AttributeName:
        print Usage
    else:

        if Options.InputFileName:
            FileName =  re.sub('\.sym', '', Options.InputFileName) + ".sym"
            if os.path.isfile(FileName):
                Options.InputFiles.append(FileName)
        else:
            for File in Options.CmdLineFileList:
                FileName =  re.sub('\.sym', '', File) + ".sym"
                if os.path.isfile(FileName):
                    Options.InputFiles.append(FileName)

        InputFiles = Options.InputFiles

        files = []
        if len(InputFiles) == 0:
            # Get list of file OurSelf
            foundOurSelf = True
            if Options.RecursiveMode:
                files = RecursiveGlob(os.getcwd(), '/*.sym')
            else:
                files = glob.glob('*.sym')

        if len(files) > 0:
            foundfiles=True
            for file in files:
                InputFiles.append(file)

        if foundOurSelf and len(InputFiles) == 0:
            QMessage (Options, 'No symbols found or specifed, try using --help')
            Options.ExitCode = 1
        elif not foundfiles :
            if len(InputFiles) > 1:
                for file in InputFiles:
                    if not os.path.isfile(file):
                        QMessage (Options, "Bad file name or option: " + file)
                        print Usage
                        Options.ExitCode = 1

        for symbol_file in InputFiles:
            ProcessSymbol(Options, symbol_file)

    exit(Options.ExitCode)

if __name__ == "__main__":
   main(sys.argv[1:])

# Done!

